core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/PathTrackingInputFormat.java

```diff
@@ -124,6 +140,103 @@ public StructuredRecord getCurrentValue() throws IOException, InterruptedExcepti
     }
 
     @Override
+    public boolean nextKeyValue() throws IOException, InterruptedException {
+      return delegate.nextKeyValue();
+    }
+
+    @Override
+    public float getProgress() throws IOException, InterruptedException {
+      return delegate.getProgress();
+    }
+
+    @Override
+    public void close() throws IOException {
+      delegate.close();
+    }
+  }
+
+  private static class TrackingAvroRecordReader extends RecordReader<NullWritable, StructuredRecord> {
+    private final RecordReader<AvroKey<GenericRecord>, NullWritable> delegate;
+    private final Schema schema;
+    private final String pathField;
+    private final String path;
+
+    private TrackingAvroRecordReader(RecordReader<AvroKey<GenericRecord>, NullWritable> delegate,
+                                     @Nullable String pathField, String path) {
+      this.delegate = delegate;
+      this.pathField = pathField;
+      this.path = path;
+      schema = getOutputSchema(pathField);
+    }
+
+    @Override
+    public void initialize(InputSplit split, TaskAttemptContext context) throws IOException, InterruptedException {
+      delegate.initialize(split, context);
+    }
+
+    @Override
+    public NullWritable getCurrentKey() throws IOException, InterruptedException {
+      return NullWritable.get();
+    }
+
+    @Override
+    public StructuredRecord getCurrentValue() throws IOException, InterruptedException {
+      AvroToStructuredTransformer recordTransformer = new AvroToStructuredTransformer();
```

Comment:
don't create a transformer for each record. Instead, create it once in initialize().